update to use Into<String> rather than &str everywhere

ChrisMacNaughton opened the pull request.

ChrisMacNaughton commented:
@emk @compressed any thoughts on this?

compressed commented:
I had actually thought about this as well. But I ran into at least two issues:

1. For functions like `fn set_secret<S: Into<String>>(&self, key: S, value: S)`, by setting up `S` like this, we restrict the user from having two different types for key and value. For instance, `key` is `&'static str` and value is a `&'a str`. Rust will expect `value` to be a `&'static str` too. I guess we would avoid this by adding additional generics for each function.
2. Some extra allocations. In that same function, we are making a `key` into a `String` when we don't actually need to (`format!` will take a `&str` just fine).

I think the idea is the right one, ideally the API should be unified to accept all "String-like" values and do the right thing, just not sure this is the right approach.

compressed commented:
There's also [Cow](https://doc.rust-lang.org/std/borrow/enum.Cow.html) which could maybe be leveraged to improve the API.

emk commented:
`fn set_secret<S: Into<String>>(&self, key: S, value: S)` should be `fn set_secret<S1: Into<String>, S2: Into<String>>(&self, key: S1, value: S2)`.

ChrisMacNaughton committed:
Replace uses of `&str` in types

As much as possible, this commit removes the use
of `&str` in function signatures and replaces it
with `Into<String>` or `AsRef<str>` depending on
what we do with the type. I am trying out the use
of `AsRef` to reduce needless copies

ChrisMacNaughton commented:
@compressed and @emk I've put a bit more thought into this and have tried to reduce the extra strings as well :)

ChrisMacNaughton committed:
Update doc tests as required

compressed commented on src/client/mod.rs:
```diff
@@ -662,7 +665,7 @@ impl<T> VaultClient<T>
     ///   .ttl(client::VaultDuration::minutes(5));
     /// let res = client.create_token(&opts).unwrap();
-    /// let mut new_client = Client::new(host, &res.client_token).unwrap();
+    /// let mut new_client = Client::new(host, &res.client_token[..]).unwrap();
```
This can probably be like on line 764:

```rust
    /// # let new_client = Client::new(host, res.client_token).unwrap();
```

ChrisMacNaughton replied:
passing ownership or a pointer ;-) I'll update it to pass ownership of the token

compressed commented on src/client/mod.rs:
```diff
@@ -612,7 +616,7 @@ impl<T> VaultClient<T>
     /// [token]: https://www.vaultproject.io/docs/auth/token.html
     pub fn renew(&mut self) -> Result<()> {
-        let res = try!(self.post("/v1/auth/token/renew-self", None, None));
+        let res = try!(self.post::<_, String>("/v1/auth/token/renew-self", None, None));
```
Annoying to have to do these turbofish everywhere... Too bad it can't figure out some type for `None`. I've ran into this many times too, I don't see any way to avoid it through.

ChrisMacNaughton replied:
I _think_ that the only places the turbofish are necessary is inside of the library, users shouldn't have to use them (I hope)

compressed reviewed (changes requested):
Sorry for the delay in reviewing!

This looks really good to me. I tested it on my existing code and didn't run into any issues like before so that's a good sign.

We could also probably bump `chrono` to `0.4` too (I tried locally, no issues that I ran into).

ChrisMacNaughton commented:
I'll make a separate PR with crate updates (incl. chrono)

ChrisMacNaughton committed:
remove explicit Deref
